### PR TITLE
Add math-block notation

### PR DESCRIPTION
**Pull request by heavywatal**
As discussed in #189, the following notations are added to math-block notation:

```tex
\[
  a = \sum \frac 1 i
\]

\begin{equation}
  a = \sum \frac 1 i
\end{equation}

\begin{equation*}
  a = \sum \frac 1 i
\end{equation*}
```

### CODE REVIEW

**Comment by burodepeper** [comment]
Cool!

As I feared, the CI tests failed. Are you familiar with that? I'd like to have a couple new ones added for these new math notations as well.

**Comment by heavywatal** [comment]
Now `\\[ ... \\]` and `\begin{equation} ... \end{equation}` are properly recognized. (I noticed that one more backslash is needed in Markdown so that it is translated into `\[ ... \]` in HTML.)

I think `\[` in HTML blocks should be matched as well because wrapping equations in `<div>` is a [common work-around to prevent underscores from being interpreted by Markdown](https://gohugo.io/content-management/formats/#mathjax-with-hugo):
```html
<div>\[
a = \sum_{i=1} \frac 1 i
\]</div>
```
Where should I add the pattern for this? maybe languege-html package, or here?

**Comment by burodepeper** [comment]
Thanks for your efforts so far! I'm a little busy at the moment, so I haven't had time to review your PR in depth yet. From what I could see, you haven't added any spec that tests for situations that it shouldn't match. Could you add a couple of those? Think about situations where a similar notation might be used, or the good notation with a typo, etc. This will help cover any regressions in the future.

The pattern you mentioned should go into language-html. Although I'm not sure why it is needed. Rather, it seems that it shouldn't be needed. Maybe the order in which the patterns are matched might also solve this issue. If it does, be sure to leave a comment describing so.

**Comment by heavywatal** [comment]
Please take your time. I am happy with using my branch on Atom for the time being.

I guess the matching order is not the only problem here because the required number of backslashes is different: `\\[` in Markdown, `\[` in HTML. But I agree that this should be in language-html.

File: grammars/repositories/flavors/math-block.cson [hunk 1]
```diff
@@ -1,13 +1,35 @@
 key: 'math-block'
 patterns: [
+  {
+    name: 'block.math.markup.md'
+    begin: '(\\${2})'
+    end: '(\\${2})(?:.*)'
```
Comment (kylebarron) [inline grammars/repositories/flavors/math-block.cson]: `(?:.*)` is incorrect. It should be just
```cson
end: '(\\${2})'
```

Pandoc, at least, allows Markdown to be on the same line as the closing `$$`. I.e.:
```md
# testing

$$
a = \sum \frac 1 i
$$ *markdown on same line*
```

is converted into: 
image

File: grammars/repositories/flavors/math-block.cson [hunk 2]
```diff
@@ -1,13 +1,35 @@
...
+  {
+    name: 'block.math.markup.md'
+    begin: '\\\\\\['
+    end: '\\\\\\]'
+    patterns: [{ include: 'text.tex.latex' }]
+    beginCaptures:
+      1: name: 'punctuation.md'
+    endCaptures:
+      1: name: 'punctuation.md'
+  }
```
Comment (kylebarron) [inline grammars/repositories/flavors/math-block.cson]: This is the correct number of backslashes, four `\` to represent a single literal `\`, and then two `\` to escape the `[`.

If it's not showing up correctly, it's the fault of language-latex and their use of `$base`. See https://github.com/burodepeper/language-markdown/pull/226, https://github.com/area/language-latex/pull/185.

On my branch of language-latex with the changes mentioned in https://github.com/area/language-latex/pull/185, I see:

image

**Review by kylebarron — CHANGES_REQUESTED** [review]

File: grammars/repositories/flavors/math-block.cson (hunk 1 above)
Comment (heavywatal) [reply grammars/repositories/flavors/math-block.cson]: Thank you for the suggestion, and sorry for my late reaction. I have just fixed it accordingly.

**Comment by kylebarron** [comment]
I think the code looks fine but it depends on https://github.com/area/language-latex/pull/185. Otherwise the latex highlighting will be off.

**Comment by heavywatal** [comment]
I may be a bit confused. Given that we need to write `\\[` in Markdown to generate `\[` in HTML, I am happy with what I get from this PR as shown in the following figure:
screen shot 2018-07-23 at 22 51 49
Are you trying to make `\[` work in Markdown by the PR 185?

**Comment by kylebarron** [comment]
> Given that we need to write \\[ in Markdown to generate \[ in HTML

What flavor of Markdown are you using?

By default Pandoc supports [only `$...$` for inline math and `$$...$$` for display math](https://pandoc.org/MANUAL.html#math). 

There are [non-Pandoc extensions](https://pandoc.org/MANUAL.html#non-pandoc-extensions) for both `\[ ... \]` and `\\[...\\]`.

> #### Extension: `tex_math_single_backslash` ####
>
> Causes anything between `\(` and `\)` to be interpreted as inline
TeX math, and anything between `\[` and `\]` to be interpreted
as display TeX math.  Note: a drawback of this extension is that
it precludes escaping `(` and `[`.
>
> #### Extension: `tex_math_double_backslash` ####
>
> Causes anything between `\\(` and `\\)` to be interpreted as inline
TeX math, and anything between `\\[` and `\\]` to be interpreted
as display TeX math.

I'm surprised to see that there's actually no definition for math in the Commonmark spec.

So which of the following should we color as math environments?

- `$...$`
- `$$...$$`
- `\(...\)`
- `\[...\]`
- `\\(...\\)`
- `\\[...\\]`

**Comment by heavywatal** [comment]
Oh, I didn't know that pandoc has such options. How does pandoc process equations? I am using Hugo/Blackfriday for HTML generation, and MathJax for equation rendering. Hence for me it is not surprising that CommonMark does not have math specs. It is MathJax's job. What Markdown has to do is generating proper HTML code to be interpreted by JavaScript. From that point of view, `tex_math_single_backslash` is inconsistent with CommonMark.

According to `short-math-guide.pdf` from AMS-Math (see below), `\begin{equation} ... \end{equation}` and `\begin{equation*} ... \end{equation*}` should also be interpreted as equations. Of course MathJax supports it. But I am not sure if it should start with single- or double-backslash in this case because either works well with Hugo.

In summary, the following patterns in Markdown should be colored IMO:
```
$...$
$$...$$
\\(...\\)
\\[...\\]
\begin{equation} ... \end{equation}
\begin{equation*} ... \end{equation*}
```

Using `$$ ... $$` is not recommended in `short-math-guide.pdf`, and I don't use that. But some people might want it...?

screen shot 2018-07-24 at 00 17 46

**Comment by kylebarron** [comment]
> How does pandoc process equations?

If converting from markdown to HTML, it converts to the appropriate HTML and uses either MathJax or KaTeX. If converting to PDF/LaTeX, it converts to the appropriate LaTeX commands and then optionally compiles the LaTeX to PDF.

I'm in agreement for
```
$...$
$$...$$
\\(...\\)
\\[...\\]
```
Supporting `$$...$$` is a must because that's the standard Pandoc way to write display math.

I'm also in agreement for the `begin`-`end` clauses, but I'm curious why you suggest those. Does Blackfriday implement those at all? I support those because Pandoc supports [Raw LaTeX input](https://pandoc.org/MANUAL.html#raw-htmltex). However if we're going to support `\begin{equation} ... \end{equation}`, we need to also support other environments from AMSMath. For example, the `align` environment is used to vertically align several equations in a row, but an error is produced if you try to nest `align` within `equation` and the practice [is not recommended](https://tex.stackexchange.com/a/372388). 

For example, this raises an error with `pandoc test.md -o test.pdf -s`:
```
# testing

\begin{equation}
\begin{align}
a = \sum \frac 1 i
\end{align}
\end{equation}
```

So if we want to support `\begin{equation} ... \end{equation}`, as I believe we should, then we should support a something like
```cson
{
begin: '\\begin\\{(align|equation|multline|split|gather|alignat|aligned|gathered|eqnarray|array|tabular)(\\*)?\\}'
end: '\\end\\{$1\\}'
}
```

As an aside, my attitude about Commonmark and Math is [this](https://talk.commonmark.org/t/mathematics-extension/457/21):
> I strongly support inclusion of Pandoc-style inline math notation, with at least the status of “you don’t have to do anything with this, but you must still parse it correctly”. For instance, in
>
>```
>Let $y = m * x + b$ where $b$ is US$50,000
>```
>
> the asterisk in m*x should not be interpreted as an emphasis marker, and conversely, the dollar sign in US$50,000 should not be interpreted as a math shift.

**Comment by kylebarron** [comment]
I didn't take the time to see which of those were still used. 

It's probably easier to take a few minutes and include a good list in this PR.

**Comment by heavywatal** [comment]
OK, I have just added simple ones: `align|equation|multline|split|gather`. It is consistent with [language-latex](https://github.com/area/language-latex/blob/2447b74978e2584e610e14d956c1c41cc1b8c7ab/grammars/latex.cson#L1424).

I excluded `alignat|aligned|gathered` because they seem to be complicated and require one more argument like `\begin{alignat}{2}`.

**Comment by kylebarron** [comment]
Alternatively, similar to how we include all rules from HTML, we could include all rules from latex after the Markdown rules. But I'm guessing that would have more side effects

**Comment by heavywatal** [comment]
I like the idea, and it seems to be a powerful and beautiful solution, but far beyond my ability and resource.

**Comment by kylebarron** [comment]
No I mean instead of writing them all ourselves, as long as the user has language-latex installed, we can just write
```cson
{include: 'text.tex.latex'}
```
And all of the latex highlighting rules get applied after the Markdown ones (depending on where that line is placed)

**Comment by heavywatal** [comment]
Yes, I think I see your point. I mean considering the side effects of including whole language-latex functions will be much harder than keeping math block grammars by ourselves. LaTeX is not just about math equations, but here we should limit the scope to math blocks.
